### ISSUE TITLE
Deploy: 2020-01-30

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#1163]
This is a fairly big deploy, so I'm going to keep some notes of the approach and any hiccups I run into.

The most complicated part of this is doing the DB migrations, so I'm going to start by doing them on a restored backup in AWS. These are pretty easy to do! I just grabbed a snapshot, restored it as a fresh DB, and made sure it was publicly accessible. Our firewall makes sure nobody can access it except from our IP. I'm going to get in there, try the migrations, and see how they go.

**Comment by mlissner** [comment]
OK, ran the migration on the demo machine. No big problem there, but I regret not turning on `\timing` first:

```
courtlistener=> BEGIN;
BEGIN
courtlistener=> --
courtlistener=> -- Remove field federal_cite_one from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "federal_cite_one" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field federal_cite_three from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "federal_cite_three" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field federal_cite_two from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "federal_cite_two" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field lexis_cite from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "lexis_cite" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field neutral_cite from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "neutral_cite" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field scotus_early_cite from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "scotus_early_cite" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field specialty_cite_one from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "specialty_cite_one" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field state_cite_one from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "state_cite_one" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field state_cite_regional from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "state_cite_regional" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field state_cite_three from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "state_cite_three" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field state_cite_two from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "state_cite_two" CASCADE;
ALTER TABLE
courtlistener=> --
courtlistener=> -- Remove field westlaw_cite from opinioncluster
courtlistener=> --
courtlistener=> ALTER TABLE "search_opinioncluster" DROP COLUMN "westlaw_cite" CASCADE;
ALTER TABLE
courtlistener=> COMMIT;
COMMIT
courtlistener=> -- Create the column without a default value
courtlistener=> ALTER TABLE "search_opinionscited" ADD COLUMN "depth" integer;
ALTER TABLE
courtlistener=> 
courtlistener=> -- Set the value to the default
courtlistener=> UPDATE "search_opinionscited" set "depth" = 1;
UPDATE 31056586
courtlistener=> 
courtlistener=> -- Make it not nullable (this is safe because our Python code will be always setting the value to the default
courtlistener=> ALTER TABLE "search_opinionscited" ALTER COLUMN "depth" SET NOT NULL;
ALTER TABLE
courtlistener=> 
courtlistener=> -- Add the index, but do it concurrently. Note that concurrent ones can't be in a transaction.
courtlistener=> CREATE INDEX CONCURRENTLY "search_opinionscited_depth_46bacaef" ON "search_opinionscited" ("depth");
CREATE INDEX
courtlistener=> BEGIN;
BEGIN
courtlistener=> --
courtlistener=> -- Alter field docket_number on docket
courtlistener=> --
courtlistener=> ALTER TABLE "search_docket" ALTER COLUMN "docket_number" TYPE text USING "docket_number"::text;
ALTER TABLE
courtlistener=> COMMIT;
COMMIT
```

The UPDATE command took a while, maybe 30 minutes. That's not great, but no locking during that; it's just an UPDATE command after all. The rest were relatively fast.

**Comment by mlissner** [comment]
I ran a test for dropping columns over in #1164. Turns out that drops should happen on the publisher first.

We have three migrations here:

1. Drop legacy columns - Start at `master`, then move down chain.

1. Add citation depth column - Start at leaf nodes, but run UPDATE on `master` once column is everywhere.

1. Convert docket number to text column - Do in any order; doesn't affect tuples.

Here we go.

**Comment by mlissner** [comment]
OK, dropping legacy columns done, no sweat. They went through almost instantly with no issues at all.

**Comment by mlissner** [comment]
OK, added column in step 2. Now updating it on master to have all values set to 1. This takes a while.

**Comment by mlissner** [comment]
One lesson for later, running a query like this and refreshing the Veteran's Legal Services page is a good indicator of whether replication is working:

    select view_count from search_docket where id=4214664;

**Comment by mlissner** [comment]
I was able to UPDATE all the values on the `master` DB to 1 on the citation table and now it's making its way to the leaf nodes. In the meantime, I ran the third migration:

```
BEGIN;
--
-- Alter field docket_number on docket
--
ALTER TABLE "search_docket" ALTER COLUMN "docket_number" TYPE text USING "docket_number"::text;
COMMIT;
```

No problems there.

**Comment by mlissner** [comment]
The update continued taking a long time, and was causing some timeouts as it made its way to from `master` to our old server. The issue there was that `master` had the default `wal_sender_timeout` parameter of 60s, so if the subscriber took a really long time to process some data, it would timeout and things would have to start over. I tweaked it in `/etc`, then reloaded the settings with:

    sudo service postgresql reload

And then things flushed through from master all the way down. This took *a while*, so I went and ate dinner. This table is the longest one we have in the system. Currently 31,058,246 rows.

As mentioned above, while that was going, I did the docket_number tweak. No problems there. 

Once the UPDATE flushed to all machines, I ran the last two alter commands to make the `depth` column not nullable and to add an index to it. Again, no problems and they completed quickly. 

In my second comment above, I didn't mention how I set up the test. I did that one by restoring an AWS backup to a new machine. That takes about five minutes, and allows us to do performance testing. Pretty great trick.

All in all, pretty smooth, though I hate how difficult these migrations are. I keep lots of notes, but nevertheless they take a lot of tests and logic to pull off every time. Postgresql needs to do something to document this or something. It's bonkers how complex it is and best practices are definitely not available yet.
